File: templates/problems/09_heap/07_find_median_stream/python/solution.py
import heapq


class SolveMedianFinder:
    """Find median in a stream using two heaps.

    AddNum: O(log n), FindMedian: O(1), Space: O(n).
    """

    def __init__(self) -> None:
        self._lo: list[int] = []  # max-heap (negated) for lower half
        self._hi: list[int] = []  # min-heap for upper half

    def add_num(self, num: int) -> None:
        heapq.heappush(self._lo, -num)
        heapq.heappush(self._hi, -heapq.heappop(self._lo))
        if len(self._hi) > len(self._lo):
            heapq.heappush(self._lo, -heapq.heappop(self._hi))

    def find_median(self) -> float:
        if len(self._lo) > len(self._hi):
            return float(-self._lo[0])
        return (-self._lo[0] + self._hi[0]) / 2.0

File: templates/problems/09_heap/07_find_median_stream/go/solution.go
package find_median_stream

import "container/heap"

// SolveMedianFinder uses two heaps to find median in a stream.
// AddNum: O(log n), FindMedian: O(1), Space: O(n).
type SolveMedianFinder struct {
	lo *solveMaxH // max-heap for lower half
	hi *solveMinHM // min-heap for upper half
}

func SolveConstructorMedianFinder() SolveMedianFinder {
	lo := &solveMaxH{}
	hi := &solveMinHM{}
	heap.Init(lo)
	heap.Init(hi)
	return SolveMedianFinder{lo: lo, hi: hi}
}

func (mf *SolveMedianFinder) AddNum(num int) {
	heap.Push(mf.lo, num)
	// move max of lo to hi
	heap.Push(mf.hi, heap.Pop(mf.lo))
	// balance: lo should have >= hi size
	if mf.hi.Len() > mf.lo.Len() {
		heap.Push(mf.lo, heap.Pop(mf.hi))
	}
}

func (mf *SolveMedianFinder) FindMedian() float64 {
	if mf.lo.Len() > mf.hi.Len() {
		return float64((*mf.lo)[0])
	}
	return (float64((*mf.lo)[0]) + float64((*mf.hi)[0])) / 2.0
}

type solveMaxH []int

func (h solveMaxH) Len() int            { return len(h) }
func (h solveMaxH) Less(i, j int) bool   { return h[i] > h[j] }
func (h solveMaxH) Swap(i, j int)        { h[i], h[j] = h[j], h[i] }
func (h *solveMaxH) Push(x interface{})  { *h = append(*h, x.(int)) }
func (h *solveMaxH) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

type solveMinHM []int

func (h solveMinHM) Len() int            { return len(h) }
func (h solveMinHM) Less(i, j int) bool   { return h[i] < h[j] }
func (h solveMinHM) Swap(i, j int)        { h[i], h[j] = h[j], h[i] }
func (h *solveMinHM) Push(x interface{})  { *h = append(*h, x.(int)) }
func (h *solveMinHM) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}
